### PR TITLE
Slider. Externalize view and calculations.

### PR DESCRIPTION
**Pull request by Ta4i**
This change is essential for proceeding with Range Slide. With this change we could reuse as much things between Simple and Range sliders as possible.

### CODE REVIEW

File: src/components/slider/slider-calculations.ts [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+
+    const relativeValue = (normilizedValue * 100) / normilizedMax;
+
+    return getValueInRange(relativeValue, 0, 100);
+}
+
+export function getAbsoluteValue(relativeValue: number, min: number, max: number): number {
+    const range = max! - min!;
+    const absoluteValue = range * relativeValue / 100 + min!;
+    return getValueInRange(absoluteValue, min!, max!);
```
Comment (wtfil) [inline src/components/slider/slider-calculations.ts]: at this point no need to add `!`. All arguments are defined here

File: src/components/slider/slider-view.tsx [hunk 2]
```diff
@@ -0,0 +1,286 @@
...
+                onTouchStart={this.onSliderAreaTouchStart}
+
+                style-state={{
+                    'active': this.props.active,
+                    'disabled': Boolean(this.props.disabled),
+                    'error': Boolean(this.props.error),
+                    'x': this.props.axis === AXISES.x !== this.props.rtl,
+                    'y': this.props.axis === AXISES.y,
+                    'x-reverse': this.props.axis === AXISES.xReverse !== this.props.rtl,
+                    'y-reverse': this.props.axis === AXISES.yReverse
```
Comment (wtfil) [inline src/components/slider/slider-view.tsx]: all states now should have camelCase

File: src/components/slider/slider-view.tsx [hunk 3]
```diff
@@ -0,0 +1,286 @@
...
+            >
+                <GlobalEvent
+                    mousemove={this.onSliderAreaMouseMove}
+                    mouseup={this.onSliderAreaMouseUp}
+                    touchmove={this.onSliderAreaTouchMove}
+                    touchend={this.onSliderAreaTouchEnd}
+                    touchcancel={this.onSliderAreaTouchEnd}
+                />
+                <input
+                    className="native-input"
```
Comment (wtfil) [inline src/components/slider/slider-view.tsx]: camelCase

File: src/components/slider/slider-view.tsx [hunk 4]
```diff
@@ -0,0 +1,286 @@
...
+
+    private getHandleStyles() {
+        const isReverse = this.isReverse(this.props.axis!);
+        return isVertical(this.props.axis!) ?
+            (isReverse ?
+                {top: `${this.props.relativeValue}%`} :
+                {bottom: `${this.props.relativeValue}%`}) :
+            (isReverse ?
+                {right: `${this.props.relativeValue}%`} :
+                {left: `${this.props.relativeValue}%`});
```
Comment (tyv) [inline src/components/slider/slider-view.tsx]: you can do 

```ts
const relativeValue = `${this.props.relativeValue}%`
const direction = // calc here 'top' | 'bottom' | 'left' | 'right'
return {[direction]: relativeValue};
```

File: src/components/slider/slider-view.tsx [hunk 5]
```diff
@@ -0,0 +1,286 @@
...
+    private getMarkStyles(position: number) {
+        const vertical = isVertical(this.props.axis!);
+        return this.isReverse(this.props.axis!) ?
+            (vertical ?
+                {top: `${position}%`} :
+                {right: `${position}%`}) :
+            (vertical ?
+                {bottom: `${position}%`} :
+                {left: `${position}%`});
+    }
```
Comment (tyv) [inline src/components/slider/slider-view.tsx]: here too, you can calculate and use `{[direction]: position}`

File: src/components/slider/slider-view.tsx [hunk 6]
```diff
@@ -0,0 +1,286 @@
...
+            ));
+        }
+
+        return markElements;
+    }
+
+    private getProgressStyles() {
+        return isVertical(this.props.axis!) ?
+            {height: `${this.props.relativeValue}%`} :
+            {width: `${this.props.relativeValue}%`};
```
Comment (tyv) [inline src/components/slider/slider-view.tsx]:
```ts
return {
  [isVertical(this.props.axis!) ? 'height' : 'width']: `${this.props.relativeValue}%`}
}
``

File: src/components/slider/slider-view.tsx [hunk 7]
```diff
@@ -0,0 +1,286 @@
...
+                    className={this.getMarkClass(position)}
+                    style={this.getMarkStyles(position)}
+                />
+            ));
+        }
+
+        return markElements;
+    }
+
+    private getProgressStyles() {
```
Comment (tyv) [inline src/components/slider/slider-view.tsx]: no return type

File: src/components/slider/slider-view.tsx [hunk 8]
```diff
@@ -0,0 +1,286 @@
...
+        return isVertical(this.props.axis!) ?
+            {height: `${this.props.relativeValue}%`} :
+            {width: `${this.props.relativeValue}%`};
+    }
+
+    private getHandleStyles() {
```
Comment (tyv) [inline src/components/slider/slider-view.tsx]: no return type

File: src/components/slider/slider-view.tsx [hunk 9]
```diff
@@ -0,0 +1,286 @@
...
+        return isVertical(this.props.axis!) ?
+            (isReverse ?
+                {top: `${this.props.relativeValue}%`} :
+                {bottom: `${this.props.relativeValue}%`}) :
+            (isReverse ?
+                {right: `${this.props.relativeValue}%`} :
+                {left: `${this.props.relativeValue}%`});
+    }
+
+    private getMarkStyles(position: number) {
```
Comment (tyv) [inline src/components/slider/slider-view.tsx]: no return type

File: src/components/slider/slider-view.tsx [hunk 10]
```diff
@@ -0,0 +1,286 @@
...
+        return this.isReverse(this.props.axis!) ?
+            (vertical ?
+                {top: `${position}%`} :
+                {right: `${position}%`}) :
+            (vertical ?
+                {bottom: `${position}%`} :
+                {left: `${position}%`});
+    }
+
+    private getMarkClass(position: number) {
```
Comment (tyv) [inline src/components/slider/slider-view.tsx]: no return type

File: src/components/slider/slider-view.tsx [hunk 11]
```diff
@@ -0,0 +1,275 @@
...
+        };
+    }
+
+    private getHandleStyles(): {[key in 'top' | 'bottom' | 'right' | 'left']?: string} {
+        return {
+            [getPositionProperty(this.props.axis!, this.props.rtl)]: `${this.props.relativeValue}%`
+        };
+    }
+
+    private getMarkStyles(position: number): {[key in 'top' | 'bottom' | 'right' | 'left']?: string} {
```
Comment (tyv) [inline src/components/slider/slider-view.tsx]: 'top' | 'bottom' | 'right' | 'left' worth own type, it is used 3 times at least

File: src/components/slider/slider-view.tsx [hunk 12]
```diff
@@ -0,0 +1,275 @@
...
+                    className={this.getMarkClass(position)}
+                    style={this.getMarkStyles(position)}
+                />
+            ));
+        }
+
+        return markElements;
+    }
+
+    private getProgressStyles(): {[key in 'width' | 'height']?: string} {
```
Comment (tyv) [inline src/components/slider/slider-view.tsx]: used twice, worth own type

File: src/components/slider/slider-calculations.ts [hunk 13]
```diff
@@ -0,0 +1,77 @@
...
+        return CONTINUOUS_STEP;
+    }
+    return 100 * step / (max - min);
+}
+
+export function getRelativeValue(value: number, min: number, max: number): number {
+    const normilizedMax = max - min;
+    const normilizedValue = value - min;
+
+    const relativeValue = (normilizedValue * 100) / normilizedMax;
```
Comment (tyv) [inline src/components/slider/slider-calculations.ts]: `normilizedValue` -- typo

**Pull request merged**